server/sonar-server/src/main/java/org/sonar/server/permission/PermissionFinder.java

```diff
@@ -19,141 +19,75 @@
-    addAnyoneGroup(dtos, query);
-    List<GroupWithPermissionDto> filteredDtos = filterMembership(dtos, query);
-
-    Paging paging = forPageIndex(query.pageIndex())
-      .withPageSize(query.pageSize())
-      .andTotal(filteredDtos.size());
 
-    return pagedGroups(filteredDtos, paging);
+    List<String> logins = users.stream().map(UserDto::getLogin).collect(Collectors.toList());
+    return permissionDao.selectUserPermissionsByQuery(dbSession, dbQuery
+      .setLogins(logins)
+      .withPermissionOnly()
+      .build());
   }
 
-  private static int offset(PermissionQuery query) {
-    int pageSize = query.pageSize();
-    int pageIndex = query.pageIndex();
-    return (pageIndex - 1) * pageSize;
-  }
-
-  private static List<GroupWithPermissionDto> filterMembership(List<GroupWithPermissionDto> dtos, PermissionQuery query) {
-    return newArrayList(Iterables.filter(dtos, new GroupWithPermissionMatchQuery(query)));
-  }
+  public List<GroupDto> findGroups(DbSession dbSession, PermissionQuery.Builder dbQuery) {
+    List<String> orderedNames = permissionDao.selectGroupNamesByPermissionQuery(dbSession, dbQuery.build());
 
-  /**
-   * As the anyone group does not exists in db, it's not returned when it has not the permission.
-   * We have to manually add it at the begin of the list, if it matched the search text
-   */
-  private static void addAnyoneGroup(List<GroupWithPermissionDto> groups, PermissionQuery query) {
-    boolean hasAnyoneGroup = Iterables.any(groups, IsAnyoneGroup.INSTANCE);
-    if (!hasAnyoneGroup
-      && !GlobalPermissions.SYSTEM_ADMIN.equals(query.permission())
-      && (query.search() == null || containsIgnoreCase(DefaultGroups.ANYONE, query.search()))) {
-      groups.add(0, new GroupWithPermissionDto().setName(DefaultGroups.ANYONE));
+    List<GroupDto> groups = groupDao.selectByNames(dbSession, orderedNames);
+    if (orderedNames.contains(DefaultGroups.ANYONE)) {
+      groups.add(0, new GroupDto().setId(0L).setName(DefaultGroups.ANYONE));
     }
-  }
 
-  private static List<GroupWithPermissionDto> pagedGroups(Collection<GroupWithPermissionDto> dtos, Paging paging) {
-    List<GroupWithPermissionDto> groups = newArrayList();
-    int index = 0;
-    for (GroupWithPermissionDto dto : dtos) {
-      if (index >= paging.offset() && groups.size() < paging.pageSize()) {
-        groups.add(dto);
-      } else if (groups.size() >= paging.pageSize()) {
-        break;
-      }
-      index++;
-    }
-    return groups;
+    return Ordering.explicit(orderedNames).onResultOf(GroupDto::getName).immutableSortedCopy(groups);
   }
 
-  private static class GroupWithPermissionMatchQuery implements Predicate<GroupWithPermissionDto> {
-    private final PermissionQuery query;
-
-    public GroupWithPermissionMatchQuery(PermissionQuery query) {
-      this.query = query;
+  public List<GroupRoleDto> findGroupPermissions(DbSession dbSession, PermissionQuery.Builder dbQuery, List<GroupDto> groups) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Document this public method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)